### PR TITLE
Use GasLimitCalculator Everywhere Instead of Function<Long, Long>

### PR DESCRIPTION
**Pull request by RatanRSur**
## PR description
In #1402 , the contributor had to do a cast because of the types we used. This helps them accomplish what they intended without that.
To that end, `GasLimitCalculator` is now just an interface. This simplifies the logic in what was previously `GasLimitCalculator` but is now `TargetingGasLimitCalculator`.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/GasLimitCalculator.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.blockcreation;
+
+import java.util.function.LongUnaryOperator;
+
+public interface GasLimitCalculator extends LongUnaryOperator {
```
Comment (shemnon) [inline ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/GasLimitCalculator.java:19]: Not a fan of the name that LongUnaryOperator forces sub-classes to use (`applyAsLong`) - it's not very readable and clear as to what is being overridden.  Do we really need the syntactic sugar features that the functional comosition provides?  i.e. will we ever actually use them?

For readability I would prefer a non-extending interface with one method `calculateNextGasTarget` so it reads clearly what is going on in subclasses.

**Review by shemnon — COMMENTED** [review]
He's still going to have to do a cast to `TargetingGasLimitCalculator ` in the exact same place.  The check is to see if it is even a mutable gas calculator and then casting to the method.  We don't want to push the mutator up to the top, so the cast seems unavoidable.  Plus, the `TargetingGasLimitCalculator` in this PR doesn't have a mutator.

Perhaps the way is to replace the gasCalculator instead of mutating it.  But that would involve wiring in the controller and passing through the changes to the created components, arguably defeating DI.  It seems like a mutating object is the least bad choice here, and the cast check is hence unavoidable.

File: ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/GasLimitCalculator.java (hunk 1 above)
Comment (RatanRSur) [reply ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/GasLimitCalculator.java:19]: Non-extending interface sounds good to me 👍

**Comment by RatanRSur** [comment]
Ok, I was planning on pushing the mutator up to the top and having it be a noop but I'm open to a cast being the less-worse option.

**Comment by shemnon** [comment]
If we push the mutator to the top and in the constant version throw an error I could get on board with that.  Then the RPC would fail if we have a constant gas limit.

**Comment by RatanRSur** [comment]
Good point, that's definitely better than a noop!

**Comment by shemnon** [comment]
Is a mutator on the gas calculator going to be added?  I feel it should be in this PR not placed in #1402.

**Comment by RatanRSur** [comment]
I was going to put it in #1402 since that is the PR that introduces runtime mutation of the target so it should bring the associated methods along with it. I think it'd be better conceptual containment.

**Comment by shemnon** [comment]
Have you communicated this with the author of #1402?

**Comment by RatanRSur** [comment]
I didn't because I assumed he'd be fine with the changes but some proactive communication wouldn't hurt. I'll do so now!

**Review by shemnon — APPROVED** [review]

**Comment by terrencecooke** [comment]
Question, I see that the PR (#1409) that you merged with my PR (#1402) had made it to the latest version of the besu executable, but my PR changes has not.  How do I get my PR into the executable? Must I make changes to my PR, or do I have to submit another pull request altogether?

**Comment by RatanRSur** [comment]
This PR doesn't include your changes in #1402, I'll be making the changes to your PR directly in order to make it compatible with the new changes and then it'll be merged.

**Comment by terrencecooke** [comment]
Okay, great. If you need anything from me to help this process along, just let me know.

**Pull request merged**